src/stub.ts

```diff
@@ -0,0 +1,139 @@
+import sinon from 'sinon';
+import { NamespaceSchema, TypeSchema, Indexable } from 'webextensions-schema';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type BrowserOut = any;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type StubOut = any;
+
+type AnySchema = {
+  $import?: string;
+  functions?: TypeSchema[];
+  properties?: Indexable<TypeSchema>;
+  events?: TypeSchema[];
+};
+
+export class BrowserBuilder {
+  public sandbox: sinon.SinonSandbox;
+  public browser: BrowserOut;
+  public aliases: Map<string, string>;
+  public types: Map<string, TypeSchema>;
+
+  private namespaceName = '';
+
+  constructor(types: Map<string, TypeSchema>) {
+    this.sandbox = sinon.createSandbox();
+    this.aliases = new Map();
+    this.browser = {
+      sinonSandbox: this.sandbox,
+    };
+    this.types = new Map(types);
+
+    // Populate typenames without namespace prefix, in cases where no match is
+    // found for lookups using fully-qualified names
+    this.types.forEach((v, k, map) => {
+      const indexOfDot = k.lastIndexOf('.');
+      if (indexOfDot !== -1) {
+        const shortKey = k.substring(indexOfDot + 1);
+        if (shortKey.length > 0 && !map.has(shortKey)) {
+          map.set(shortKey, v);
+        }
+      }
+    });
+  }
+
+  public namespace(namespace: NamespaceSchema): void {
+    this.namespaceName = namespace.namespace;
```

Comment:
I'd rather move this into `private schema` so `this.namespaceName = '';` isn't needed anymore.

Reply:
`schema()` gets called recursively by `ref()` (which is called by `property()` and `fn()`).These are nested recursive calls. Only the top-level call `namespace()` has an actual `namespace` object.

The calls go as follows:

`namespace() -> schema() -> property()/fn() -> ref() -> schema() -> property()/fn() -> ref() -> schema() etc. etc.`

I didn't want to keep passing the `namespace` object from function to function (to keep the code simpler) so I had the top-level `namespace()` function set and unset it before and after the recursive loop. But let me know if you can think of a cleaner way to do it, I'll happily make the change!

Reply:
Forget my last comment, this is now fixed in: c6eb1e6